### PR TITLE
ROX-20160: Fetch external networks and load them into the main image

### PR DESCRIPTION
**Pull request by tommartensen**
## Description

Brings https://gitlab.cee.redhat.com/stackrox/rhacs-midstream/-/blob/rhacs-1.0-rhel-8/distgit/containers/rhacs-main/Dockerfile.in?ref_type=heads#L109-120 to the Konflux image.
This is the last piece of the onboarding puzzle.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

- Konflux build passing
- GHA main build passing
- Confirming that `/stackrox/static-data/docs/api/v2/swagger.json`, `/stackrox/static-data/docs/api/v1/swagger.json` and `/stackrox/static-data/external-networks/external-networks.zip` exist and have valid contents.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: .tekton/main-pull-request.yaml [hunk 1]
```diff
@@ -299,6 +299,22 @@ spec:
...
+    - name: fetch-external-networks
+      runAfter:
+      - clone-repository
+      taskSpec:
+        steps:
+          - name: fetch-external-networks
+            image: registry.access.redhat.com/ubi8-minimal:latest
+            script: |
+              #!/usr/bin/env bash
+              microdnf install -y openssl zip
```
Comment (msugakov) [inline .tekton/main-pull-request.yaml:311]: Why did you add `openssl`, is it because it brings something needed for `curl` to function?

File: .tekton/main-pull-request.yaml [hunk 2]
```diff
@@ -299,6 +299,22 @@ spec:
...
 
+    - name: fetch-external-networks
+      runAfter:
+      - clone-repository
+      taskSpec:
+        steps:
+          - name: fetch-external-networks
+            image: registry.access.redhat.com/ubi8-minimal:latest
+            script: |
+              #!/usr/bin/env bash
```
Comment (msugakov) [inline .tekton/main-pull-request.yaml:310]: Please add the usual mantra for failing-fast:
```suggestion
              #!/usr/bin/env bash
              set -euo pipefail
```
Asking because the script now is more than one-liner :-)

File: .tekton/main-pull-request.yaml [hunk 3]
```diff
@@ -299,6 +299,22 @@ spec:
...
+              microdnf install -y openssl zip
+              "$(workspaces.source.path)/source/image/rhel/fetch-stackrox-data.sh" \
+                  "$(workspaces.source.path)/source/.konflux/stackrox-data"
```
Comment (msugakov) [inline .tekton/main-pull-request.yaml:313]: [minor and optional] Since it's the last command of the script, you can `exec` into it which saves from creating a child process.
```suggestion
              exec "$(workspaces.source.path)/source/image/rhel/fetch-stackrox-data.sh" \
                  "$(workspaces.source.path)/source/.konflux/stackrox-data"
```

File: .tekton/main-pull-request.yaml [hunk 4]
```diff
@@ -299,6 +299,22 @@ spec:
           image: registry.access.redhat.com/ubi8/ubi:latest
           script: exec "$(workspaces.source.path)/source/scripts/konflux/subscription-manager/subscription-manager-bro.sh" smuggle
 
+    - name: fetch-external-networks
+      runAfter:
+      - clone-repository
+      taskSpec:
+        steps:
+          - name: fetch-external-networks
```
Comment (msugakov) [inline .tekton/main-pull-request.yaml:307]: My initial idea was to have this download as another step of the existing `fetch-external-content` task.
It is also fine to keep this download separately but in this case I'd ask you to rename `fetch-external-content` to something more specific, like `copy-subscription-manager-activation-key`.

File: image/rhel/fetch-stackrox-data.sh [hunk 5]
```diff
@@ -4,7 +4,11 @@
 
 set -euxo pipefail
 
+TARGET_DIR="${1:-/stackrox-data}"
```
Comment (msugakov) [inline image/rhel/fetch-stackrox-data.sh:7]: I like how you addressed it by making the command extensible and not breaking the existing caller. I would, however, ask to take it a bit further and convert the single parameter to be required instead of being optional. This is to make it explicit _at the call site_ where the data would go.

Besides `konflux.Dockerfile`, there's just one place where this `fetch-stackrox-data.sh` is called upstream. The script is not usable downstream so you don't need to worry about breaking anything there.

```suggestion
if [[ "$#" -lt "1" ]]; then
	>&2 echo "Error: please provide target directory where to store downloaded data"
	exit 6
fi

TARGET_DIR="${1}"
```

(untested)

File: image/rhel/fetch-stackrox-data.sh [hunk 6]
```diff
@@ -4,7 +4,11 @@
...
+
 fetch_stackrox_data() {
+    local target_dir="${1}"
```
Comment (msugakov) [inline image/rhel/fetch-stackrox-data.sh:10]: [nit] `TARGET_DIR` and `target_dir` mean the same, maybe the former is just redundant since it's used in only one place? I suggest to call `fetch_stackrox_data "${1}"` instead of `fetch_stackrox_data "${TARGET_DIR}"`.

File: image/rhel/fetch-stackrox-data.sh [hunk 7]
```diff
@@ -4,7 +4,11 @@
...
+
     mkdir -p /tmp/external-networks
```
Comment (msugakov) [inline image/rhel/fetch-stackrox-data.sh:12]: I think I may have reviewed this script but now the amount of times `/tmp/external-networks` is repeated worries me. If you don't mind improving things more, I suggest using `mktemp` and environment variable instead.

```suggestion
    local download_dir
    download_dir="$(mktemp -d --tmpdir external-networks.XXXXXXXXXX)"
```

`--tmpdir` and `external-networks.XXXXXXXXXX` pattern are optional cosmetics (in case they aren't available on all platforms).

File: image/rhel/fetch-stackrox-data.sh [hunk 8]
```diff
@@ -16,9 +20,9 @@ fetch_stackrox_data() {
 
     sha256sum -c <( echo "$(cat /tmp/external-networks/checksum) /tmp/external-networks/networks" )
 
-    mkdir /stackrox-data/external-networks
-    zip -jr /stackrox-data/external-networks/external-networks.zip /tmp/external-networks
+    mkdir -p "${target_dir}/external-networks"
+    zip -jr "${target_dir}/external-networks/external-networks.zip" /tmp/external-networks
```
Comment (msugakov) [inline image/rhel/fetch-stackrox-data.sh:24]: I don't know why we have `--test` option in `zip` command downstream and if it ever saved us from anything but it looks like something which wouldn't hurt having.
```suggestion
    zip -jr --test "${target_dir}/external-networks/external-networks.zip" /tmp/external-networks
```

File: image/rhel/fetch-stackrox-data.sh (hunk 7 above)
Comment (tommartensen) [reply image/rhel/fetch-stackrox-data.sh]: Done

File: image/rhel/fetch-stackrox-data.sh (hunk 6 above)
Comment (tommartensen) [reply image/rhel/fetch-stackrox-data.sh:13]: Done.

File: image/rhel/fetch-stackrox-data.sh (hunk 5 above)
Comment (tommartensen) [reply image/rhel/fetch-stackrox-data.sh]: It works!

File: .tekton/main-pull-request.yaml (hunk 4 above)
Comment (tommartensen) [reply .tekton/main-pull-request.yaml:307]: Done in https://github.com/stackrox/stackrox/pull/10672/commits/25f21058a2586f79b61718be2ff9b6229e91852d

File: .tekton/main-pull-request.yaml (hunk 1 above)
Comment (tommartensen) [reply .tekton/main-pull-request.yaml]: I added it because it was present in the upstream Dockerfile: https://github.com/stackrox/stackrox/blob/master/image/rhel/Dockerfile#L19. 

However, it is not required and I removed it in https://github.com/stackrox/stackrox/pull/10672/commits/fdb6d014a9bd20467448de2dd0565edf0ecad37d

**Review by msugakov — APPROVED** [review]
